File: solutions/python/binary_tree_level_order.py
# Binary_tree_level_order.cpp 100b4adabfd008775520062bd407c3323ea646af
import collections
from binary_tree_prototype import BinaryTreeNode


results = []
one_line_result = []


# @include
def print_binary_tree_depth_order(tree):
    processing_nodes = collections.deque()
    processing_nodes.append(tree)
    num_nodes_current_level = len(processing_nodes)
    while processing_nodes:
        curr = processing_nodes.popleft()
        num_nodes_current_level -= 1
        if not curr:
            continue
        print(curr.data, end=' ')
        # @exclude
        one_line_result.append(curr.data)
        # @include

        # Defer the null checks to the null test above.
        processing_nodes.append(curr.left)
        processing_nodes.append(curr.right)
        # Done with the nodes at the current depth.
        if num_nodes_current_level == 0:
            print()
            num_nodes_current_level = len(processing_nodes)
            # @exclude
            results.append(one_line_result.copy())
            one_line_result.clear()
            # @include
# @exclude


def main():
    #      3
    #    2   5
    #  1    4 6
    tree = BinaryTreeNode(3)
    tree.left = BinaryTreeNode(2)
    tree.left.left = BinaryTreeNode(1)
    tree.right = BinaryTreeNode(5)
    tree.right.left = BinaryTreeNode(4)
    tree.right.right = BinaryTreeNode(6)
    # should output 3
    #               2 5
    #               1 4 6
    print_binary_tree_depth_order(tree)
    assert results == [[3], [2, 5], [1, 4, 6]]


if __name__ == '__main__':
    main()

File: solutions/cpp/binary_tree_level_order.cc
// Copyright (c) 2015 Elements of Programming Interviews. All rights reserved.

#include <cassert>
#include <iostream>
#include <memory>
#include <queue>
#include <vector>

#include "./Binary_tree_prototype.h"

using std::cout;
using std::endl;
using std::equal;
using std::make_unique;
using std::move;
using std::queue;
using std::unique_ptr;
using std::vector;

// @include
vector<vector<int>> BinaryTreeDepthOrder(
    const unique_ptr<BinaryTreeNode<int>>& tree) {
  queue<BinaryTreeNode<int>*> curr_depth_nodes({tree.get()});
  vector<vector<int>> result;

  while (!curr_depth_nodes.empty()) {
    queue<BinaryTreeNode<int>*> next_depth_nodes;
    vector<int> this_level;
    while (!curr_depth_nodes.empty()) {
      auto curr = curr_depth_nodes.front();
      curr_depth_nodes.pop();
      if (curr) {
        this_level.emplace_back(curr->data);

        // Defer the null checks to the null test above.
        next_depth_nodes.emplace(curr->left.get());
        next_depth_nodes.emplace(curr->right.get());
      }
    }

    if (!this_level.empty()) {
      result.emplace_back(this_level);
    }
    curr_depth_nodes = next_depth_nodes;
  }
  return result;
}
// @exclude

int main(int argc, char* argv[]) {
  //      3
  //    2   5
  //  1    4 6
  // 10
  // 13
  unique_ptr<BinaryTreeNode<int>> tree = make_unique<BinaryTreeNode<int>>(
      BinaryTreeNode<int>{3, nullptr, nullptr});
  tree->left = make_unique<BinaryTreeNode<int>>(
      BinaryTreeNode<int>{2, nullptr, nullptr});
  tree->left->left = make_unique<BinaryTreeNode<int>>(
      BinaryTreeNode<int>{1, nullptr, nullptr});
  tree->left->left->left = make_unique<BinaryTreeNode<int>>(
      BinaryTreeNode<int>{10, nullptr, nullptr});
  tree->left->left->left->right = make_unique<BinaryTreeNode<int>>(
      BinaryTreeNode<int>{13, nullptr, nullptr});
  tree->right = make_unique<BinaryTreeNode<int>>(
      BinaryTreeNode<int>{5, nullptr, nullptr});
  tree->right->left = make_unique<BinaryTreeNode<int>>(
      BinaryTreeNode<int>{4, nullptr, nullptr});
  tree->right->right = make_unique<BinaryTreeNode<int>>(
      BinaryTreeNode<int>{6, nullptr, nullptr});
  // should output 3
  //               2 5
  //               1 4 6
  //               10
  //               13
  auto result = BinaryTreeDepthOrder(tree);
  vector<vector<int>> golden_res = {{3}, {2, 5}, {1, 4, 6}, {10}, {13}};
  assert(equal(golden_res.begin(), golden_res.end(), result.begin(),
               result.end()));
  return 0;
}
